### ISSUE TITLE
conch and getPeer

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2997]
|<img alt="bernat@...'s avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| bernat@... reported|
|-|-|
|Trac ID|trac#2997|
|Type|defect|
|Created|2008-01-14 17:38:00Z|
|Branch|https://github.com/twisted/twisted/tree/conch-session-getPeer-2997|

I would like to use getPeer from conch but there is no easy way of this. For example, if I take ```sshsimpleserver.py``` and apply the following patch:
 ```
#!diff
--- sshsimpleserver.py~	2008-01-14 18:31:43.000000000 +0100
+++ sshsimpleserver.py	2008-01-14 18:32:13.000000000 +0100
@@ -30,6 +30,7 @@
     """this is our example protocol that we will run over SSH
     """
     def dataReceived(self, data):
+        print self.transport.getPeer()
         if data == '\r':
             data = '\r\n'
         elif data == '\x03': #^C
```

I get:
 ```python
	exceptions.AttributeError: SSHSessionProcessProtocol instance has no attribute 'getPeer'
```

The solution seems to be to use ```self.transport.session.getPeer()```. I think that getPeer method should be implemented in ```SSHSessionProcessProtocol``` (to be able to use a protocol with SSH or without SSH for example).

In fact, there seems to be a bug in ```SSHChannel```. When using ```self.transport.session.getPeer()```, I get :
 ```python
	  File "/usr/lib/python2.4/site-packages/twisted/conch/ssh/channel.py", line 219, in getPeer
	    return('SSH', )+self.conn.transport.getPeer()
	exceptions.AttributeError: SSHServerTransport instance has no attribute 'getPeer'
```

The fix should be to use ```self.conn.transport.transport.getPeer()```.

<details><summary>Searchable metadata</summary>

```
trac-id__2997 2997
type__defect defect
reporter__bernat____ bernat@...
priority__normal normal
milestone__ 
branch__branches_conch_session_getPeer_2997 branches/conch-session-getPeer-2997
branch_author__z3p z3p
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1200332280000000 1200332280000000
changetime__1338057269000000 1338057269000000
version__None None
owner__z3p z3p

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="bernat@...'s avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|bernat@... commented|
|-|-|

The first part of this bug is in fact already in [#2453](https://github.com/twisted/twisted/issues/2453). Sorry for that. The second part is still valid though.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|z3p commented|
|-|-|

(In [33133]) Branching to 'conch-session-getPeer-2997'

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|z3p commented|
|-|-|

Ready for review.

Buildbot results: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/conch-session-getPeer-2997

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|z3p commented|
|-|-|

Forgot the review keyword.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|z3p removed owner|
|-|-|

I shouldn't review my own ticket.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst set owner to z3p|
|-|-|

Glad you're back to working on conch!

Major issues:
   1. The tests looked like they failed on the Windows build slaves.

Minor Issues:
   1. You're doing (in address.py) `'SSHTransportAddress(%r)' % self.address`. Better to always do `'SSHTransportAddress(%r)' % (self.address,)` in case some jerk decides to pass in a tuple.

Coding standard issues:
   1. You're missing a news file in `twisted/conch/topfiles/`.
   1. The new address test class should have a docstring.
   1. Two lines between each method in all the code you added, rather than one.
   1. In test docstrings you're using `L{code}` to refer to objects that don't exist at the given import path; L is supposed to be for links, to objects that are actually in that python module at that full or relative Python path. Use `C{}` for arbitrary code snippets and references. In practice this was in tests, where docstring correctness doesn't matter.
   1. Some lines in transport.py are longer than 80 chars.
   1. `@ivar` secondary lines (`address.py`) are usually indented, though I have no idea if that's actually required.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @teratorn|
|-|-|

[#5566](https://github.com/twisted/twisted/issues/5566) was a duplicate of this.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|z3p removed owner|
|-|-|

I believe I've addressed those review comments in the latest commit.  I'm having trouble loading BuildBot at the moment, but I did start the builds: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/conch-session-getPeer-2997

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve set owner to z3p|
|-|-|

1.Pyflakes: `twisted/conch/test/test_transport.py:573: local variable 'testCipher' is assigned to but never used`

 2. It's better to only import SSHTransportAddress instead of the whole address module if you only use that object.

 3. Please add @since marker to your new module and your 2 new methods.

 4. Please put 3 lines after the imports and before your first class.

All very minor, please merge once fixed. Thanks!

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|z3p commented|
|-|-|

(In [34443]) comment 1: pyflakes/80cols

Refs [#2997](https://github.com/twisted/twisted/issues/2997)

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|z3p commented|
|-|-|

(In [34445]) comment 4: 3 lines between imports and first class

Refs [#2997](https://github.com/twisted/twisted/issues/2997)

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|z3p set status to `closed`|
|-|-|

(In [34448]) Merge conch-session-getPeer-2997

Author: z3p
Reviewer: itamar, therve
Fixes: [#2997](https://github.com/twisted/twisted/issues/2997)

Implements getPeer()/getHost() for the SSHTransport.  This also creates a new
SSHTransportAddress which contains the data for those methods.
